### PR TITLE
Fix #930: add a `.gemini/styleguide.md` for Gemini Code Assist

### LINKED ISSUE #930: Create a `styleguide.md` for Gemini Code Assist
### What is the task?

Described in https://developers.google.com/gemini-code-assist/docs/customize-gemini-behavior-github#styleguide.md

### (Optional) Do you have ideas or preferences for the approach?

_No response_

### How urgent is this for you?

None

### PR DESCRIPTION
**Pull request by mhucka**
Add a guide for Gemini Code Assist to teach it about the conventions used in this project.

[Gemini Code Assist](https://codeassist.google/) is a coding assistant based on Google Gemini. Among other things, it has an [integration with GitHub](https://developers.google.com/gemini-code-assist/docs/set-up-code-assist-github). Google has enabled it in our organization. It acts as a [code reviewer for pull requests](https://developers.google.com/gemini-code-assist/docs/review-github-code). Its behavior can be [customized](https://developers.google.com/gemini-code-assist/docs/customize-gemini-behavior-github) to some extent, and it can be informed about the practices and conventions used in a project through the use of a [styleguide.md file](https://developers.google.com/gemini-code-assist/docs/customize-gemini-behavior-github#styleguide.md).

The `styleguide.md` file in this PR is essentially a draft contributors' guide for the qsim project. The contents can serve as a starting point for adding a section to the qsim documentation.

### CODE REVIEW

File: .gemini/styleguide.md [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+
+*   _Performance_: While readability is paramount, code should be efficient.
+
+### Overall development approach
+
+*   Use test-driven development (TDD) whenever possible. Write a failing test
+    _before_ writing new code. This ensures testability and clarifies
+    requirements.
+
+*   Isolate tests. tests must be independent and must not rely on the state of
```
Comment (pavoljuhas) [inline .gemini/styleguide.md]:
```suggestion
*   Isolate tests. Tests must be independent and must not rely on the state of
```

File: .gemini/styleguide.md [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+
+*   Use asserts intelligently. Test assertions should be specific. Instead of
+    just asserting `true`, assert that a specific value equals an expected
+    value. Provide meaningful failure messages.
+
+*   Test coverage must be high. We don't require 100% coverage, but any
+    uncovered code must be annotated with `# pragma: no cover`. To ignore
+    coverage of a single line, place `# pragma: no cover` at the end of the
+    line. To ignore coverage for an entire block, start the block with a `#
+    pragma: no cover` comment on its own line.
```
Comment (pavoljuhas) [inline .gemini/styleguide.md]: Nit - for plain-text readability
```suggestion
    line. To ignore coverage for an entire block, start the block with a
    `# pragma: no cover` comment on its own line.
```

File: .gemini/styleguide.md [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+*   [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html)
+
+*   [Google Python Style
+    Guide](https://google.github.io/styleguide/pyguide.html)
+
+*   [Google Markdown Style
+    Guide](https://google.github.io/styleguide/docguide/style.html)
+
+*   [Google Shell Style
+    Guide](https://google.github.io/styleguide/shellguide.html)
```
Comment (pavoljuhas) [inline .gemini/styleguide.md]: Nit (disregard if original line breaks were from some markdown formatter) - for a plain text readability it is better to break the line before the URL value:
```suggestion
*   [Google Python Style Guide](
    https://google.github.io/styleguide/pyguide.html)

*   [Google Markdown Style Guide](
    https://google.github.io/styleguide/docguide/style.html)

*   [Google Shell Style Guide](
    https://google.github.io/styleguide/shellguide.html)
```

Please fix other similar instances below.

File: .gemini/styleguide.md [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+
+This project uses [Google style doc
+strings](http://google.github.io/styleguide/pyguide.html#381-docstrings) with a
+Markdown flavor and support for LaTeX. Docstrings use tripe double quotes, and
+the first line should be a concise one-line summary of the function or object.
+
+Here is an example docstring:
+
+```python
+def some_method(a: int, b: str) -> float:
```
Comment (pavoljuhas) [inline .gemini/styleguide.md]: nit - maybe `some_function` is better as method is usually associated with a class:
```suggestion
def some_function(a: int, b: str) -> float:
```

File: .gemini/styleguide.md [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+
+*   _Classes_/_Structs_: Use CapWords (CamelCase). Examples: `UserManager`,
+    `PaymentProcessor`.
+
+*   _Namespaces_: Use snake_case.
+
+*   _Enum Members_: Usually prefixed with `k` followed by a CamelCase name.
+    Examples: `kX`, `kY`, as seen from Python.
+
+#### Template names
```
Comment (pavoljuhas) [inline .gemini/styleguide.md]: This and the following *`For` and `ForArgs`* sections are better suited for API docs  than for a style guide.  That text is in fact already present in [docs/type_reference.md](https://github.com/quantumlib/qsim/blob/a80369d0c3dedbc15cf0d1ab81a8e84cfc638242/docs/type_reference.md).  Human developers are unlikely to read this from a hidden `.gemini` folder.  Gemini should be able to find this information in type_reference.md or infer it from the code.

Please remove so we do not have duplicate text to maintain.

File: .gemini/styleguide.md (hunk 3 above)
Comment (mhucka) [reply .gemini/styleguide.md]: > Nit (disregard if original line breaks were from some markdown formatter) - for a plain text readability it is better to break the line before the URL value:
> 
> Please fix other similar instances below.

I agree they're more readable. Will fix in next commit.

File: .gemini/styleguide.md (hunk 5 above)
Comment (mhucka) [reply .gemini/styleguide.md]: You're right. Removed in the next commit.

**Comment by mhucka** [comment]
> Please remove the _Template names_ and _For and For Args_ sections. Otherwise LGTM with a few non-essential suggestions.

Thank you for the review and the feedback!

File: .gemini/styleguide.md [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+*   _Maintainability_: Code should be easy to modify and extend.
+
+*   _Consistency_: Adhering to a consistent style across all projects improves
+    collaboration and reduces errors.
+
+*   _Performance_: While readability is paramount, code should be efficient.
+
+### Overall development approach
+
+*   Use test-driven development (TDD) whenever possible. Write a failing test
```
Comment (sergeisakov) [inline .gemini/styleguide.md]: I think test-driven development was never used for this project.

File: .gemini/styleguide.md [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+    for each test run.
+
+*   When new functions, classes, and files are introduced, they should also have
+    corresponding tests. Existing tests must continue to pass (or be updated)
+    when changes are introduced, and code should be covered by tests.
+
+*   Make sure to cover edge cases: Test for invalid inputs, null values, empty
+    arrays, zero values, and off-by-one errors.
+
+*   Use mocking for dependencies. In unit tests, external dependencies (e.g.,
```
Comment (sergeisakov) [inline .gemini/styleguide.md]: The only external dependency of the C++ code is Eigen3. This dependence is not propagated to qsimcirq.

File: .gemini/styleguide.md [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+*   Test coverage must be high. We don't require 100% coverage, but any
+    uncovered code must be annotated with `# pragma: no cover`. To ignore
+    coverage of a single line, place `# pragma: no cover` at the end of the
+    line. To ignore coverage for an entire block, start the block with a
+    `# pragma: no cover` comment on its own line.
+
+### Overall code format conventions
+
+This project generally follows Google coding conventions, with a few changes
```
Comment (sergeisakov) [inline .gemini/styleguide.md]: Yes, this project follows Google coding conventions. However, there are some exceptions (at least for the C++ code).

File: .gemini/styleguide.md [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    proper punctuation.
+
+### Overall nomenclature conventions
+
+This project follows some nomenclature rules and conventions in order to
+maintain a consistent interface that is easy to use. By using consistent naming,
+we can reduce cognitive load on human users and developers. Please try to use
+these terms when writing code.
+
+*   Use `state_vector` to describe a pure state.  **Do not** use `wavefunction`,
```
Comment (sergeisakov) [inline .gemini/styleguide.md]: This and the two following items are Python-specific guidance. In the C++ code, `state` is used everywhere for state vectors. A computational basis state (say, |0000>) is typically referred to as `bitstring`.

File: .gemini/styleguide.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+*   A function argument (`state_vector`, `state_rep`, or `initial_state`)
+    should permit any of the possible representations of a state: A NumPy
+    array, a NumPy tensor, an integer representing a qubit-system's
+    computational basis state, a sequence of _n_ integers representing a
+    qudit's basis state, or a `cirq.ProductState`. The type annotation should
+    be `cirq.STATE_VECTOR_LIKE` and you should use `cirq.to_valid_state_vector`
+    to canonicalize as a NumPy array of amplitudes. If a function expects a
+    NumPy array of amplitudes, its type annotation should be `np.ndarray`.
+
+*   Use `density_matrix` to describe a mixed state that is passed in as a NumPy
```
Comment (sergeisakov) [inline .gemini/styleguide.md]: There are no density matrices in qsim.

**Review by sergeisakov — COMMENTED** [review]
It looks good to me in general. I just have a number of comments.

File: .gemini/styleguide.md (hunk 6 above)
Comment (mhucka) [reply .gemini/styleguide.md]: Got it. The next push will remove the text about TDD.

File: .gemini/styleguide.md (hunk 8 above)
Comment (mhucka) [reply .gemini/styleguide.md]: I tried to note this in the document, but apart from indentation and line length conventions, I'm not sure what the other differences are. @sergeisakov Could you let me know what could be mentioned as differences? It doesn't have to be exhaustive.

File: .gemini/styleguide.md (hunk 9 above)
Comment (mhucka) [reply .gemini/styleguide.md]: I removed the "overall nomenclature" section, moved these python-specific items to the "python naming conventions" section, and added a new section for C++ naming conventions with those two points in it (about state and bitstring).

This will be in the next push on this branch.

File: .gemini/styleguide.md (hunk 10 above)
Comment (mhucka) [reply .gemini/styleguide.md]: Thank you. Deleted in the next update.

File: .gemini/styleguide.md (hunk 8 above)
Comment (mhucka) [reply .gemini/styleguide.md]: In the interest of time, I'll go ahead and merge the (edited and corrected) latest version of this file. I want to add some of this coding guidance to the docs for humans. Info about the differences from Google style will be useful to put there.
